pinot-core/src/main/java/org/apache/pinot/core/transport/AsyncQueryResponse.java

```diff
@@ -56,13 +56,17 @@ public AsyncQueryResponse(QueryRouter queryRouter, long requestId, Set<ServerRou
     _requestId = requestId;
     int numServersQueried = serversQueried.size();
     _responseMap = new ConcurrentHashMap<>(HashUtil.getHashMapCapacity(numServersQueried));
+    _serverRoutingStatsManager = serverRoutingStatsManager;
     for (ServerRoutingInstance serverRoutingInstance : serversQueried) {
+      // Record stats related to query submission just before sending the request. Otherwise, if the response is
+      // received immediately, there's a possibility of updating query response stats before updating query
+      // submission stats.
+      _serverRoutingStatsManager.recordStatsAfterQuerySubmission(requestId, serverRoutingInstance.getInstanceId());
```

Comment:
Per the method name, this should be put within either `markRequestSubmitted()` or `markRequestSent()`

Reply:
For low latency usecases at high qps, this can cause the issue mentioned in comments. Hence we update the stats just before submitting. 
I've updated the method name to better convey this.